src/diverse_seq/distance.py

```diff
@@ -1,28 +1,473 @@
-from functools import singledispatch
+import heapq
+import math
+from collections.abc import Sequence
+from typing import Literal, TypeAlias
 
-from numpy import intersect1d, ndarray, union1d
+import cogent3.app.typing as c3_types
+import numpy as np
+from cogent3.app.composable import define_app
+from rich.progress import Progress
 
+from diverse_seq.record import (
+    KmerSeq,
+    SeqArray,
+    _get_canonical_states,
+    make_kmerseq,
+    seq_to_seqarray,
+)
 
-@singledispatch
-def _intersect_union(rec1, rec2) -> tuple[int, int]:
-    raise NotImplementedError
+BottomSketch: TypeAlias = list[int]
 
 
-@_intersect_union.register(set)
-def _(rec1: set, rec2: set) -> tuple[int, int]:
-    intersect = rec1 & rec2
-    union = rec1 | rec2
-    return len(intersect), len(union)
+@define_app
+class dvs_dist:
+    """return the N most divergent sequences"""
 
+    def __init__(
+        self,
+        distance_mode: Literal["mash", "euclidean"] = "mash",
+        *,
+        k: int = 16,
+        sketch_size: int | None = None,
+        moltype: str = "dna",
+        mash_canonical_kmers: bool | None = None,
+        with_progress: bool = True,
+    ) -> None:
+        """Initialise parameters for kmer distance calculation.
 
-@_intersect_union.register(ndarray)
-def _(rec1: ndarray, rec2: ndarray) -> tuple[int, int]:
-    intersect = intersect1d(rec1, rec2, assume_unique=True)
-    union = union1d(rec1.data, rec2.data)
-    return len(intersect), len(union)
+        Parameters
+        ----------
+        distance_mode : Literal[&quot;mash&quot;, &quot;euclidean&quot;], optional
```

Comment:
remove html escapes, you can use double quotes within docstrings. My pref is leave the type hints to the python code

Reply:
ah, my automatic docstring thing put them in automatically (as with the type hints - can remove them in the docstring if you'd like me to).